File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/stc/STCViewModel.kt
package com.openu60.feature.router.stc

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.STCConfig
import com.openu60.core.model.STCParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class STCState(
    val config: STCConfig = STCConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class STCViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(STCState())
    val state: StateFlow<STCState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val paramsData = agentClient.getJSON("/api/modem/stc/params")
                var config = STCParser.parseParams(paramsData)

                try {
                    val statusData = agentClient.getJSON("/api/modem/stc/status")
                    config = STCParser.parseStatus(statusData, config)
                } catch (_: Exception) {}

                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun toggle(enabled: Boolean) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/modem/stc", mapOf("stc_enable" to if (enabled) "1" else "0"))
                _state.value = _state.value.copy(message = "STC ${if (enabled) "enabled" else "disabled"}", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) toggle(enabled) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}

File: mobile/ios/OpenU60/Features/RouterSettings/STC/STCViewModel.swift
import SwiftUI

@Observable
@MainActor
final class STCViewModel {
    var config: STCConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    var editLteTimer: String = ""
    var editNrsaTimer: String = ""
    var editLteMax: String = ""
    var editNrsaMax: String = ""

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let paramsData = try await client.getJSON("/api/cell/stc/params")
            config = STCParser.parseParams(paramsData)

            if let statusData = try? await client.getJSON("/api/cell/stc/status") {
                config = STCParser.parseStatus(statusData, into: config)
            }

            editLteTimer = config.lteCollectTimer
            editNrsaTimer = config.nrsaCollectTimer
            editLteMax = config.lteWhitelistMax
            editNrsaMax = config.nrsaWhitelistMax
        } catch {
            showMessage("Failed to load STC: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func applyParams() async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/cell/stc/params", body: [
                "lte_collect_timer": editLteTimer,
                "nrsa_collect_timer": editNrsaTimer,
                "lte_whitelist_max": editLteMax,
                "nrsa_whitelist_max": editNrsaMax
            ])
            showMessage("STC parameters updated", isError: false)
            config.lteCollectTimer = editLteTimer
            config.nrsaCollectTimer = editNrsaTimer
            config.lteWhitelistMax = editLteMax
            config.nrsaWhitelistMax = editNrsaMax
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func enable() async {
        isLoading = true
        do {
            let _ = try await client.postJSON("/api/cell/stc/enable")
            showMessage("STC enabled", isError: false)
            config.enabled = true
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func disable() async {
        isLoading = true
        do {
            let _ = try await client.postJSON("/api/cell/stc/disable")
            showMessage("STC disabled", isError: false)
            config.enabled = false
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func reset() async {
        isLoading = true
        do {
            let _ = try await client.postJSON("/api/cell/stc/reset")
            showMessage("STC whitelist reset", isError: false)
            config.enabled = false
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
